backend/generator/Generator.hs

```diff
@@ -0,0 +1,110 @@
+  -- ^ How many compounds to put into one file.
+  , gsTargetNum :: Word32
+  -- ^ How many targets to put into one file.
+  , gsMeasurementsNum :: Word
+  -- ^ Number of measurements in one experiment.
+  }
+
+-- | Generate some dummy data and save it to DB according to the provided spec.
+generateAndSave :: GeneratorSpec -> Edna ()
+generateAndSave spec@GeneratorSpec {..} = do
+  initEdna
+
+  mapM_ addProject [1 .. gsProjectNum]
+  logMessage $ "Added " <> show gsProjectNum <> " projects"
+
+  mapM_ addMethodology [1 .. gsMethodologyNum]
+  logMessage $ "Added " <> show gsMethodologyNum <> " methodologies"
+
+  addExperiments spec
+  where
+    addProject i = Library.addProject ProjectReq
+      { prqName = "Project #" <> show i
+      , prqDescription = show i <$ guard (even i)
+      }
+
+    addMethodology i = Library.addMethodology MethodologyReq
+      { mrqName = "Methodology #" <> show i
+      , mrqDescription = show i <$ guard (even i)
+      , mrqConfluence = Nothing
+      }
+
+addExperiments :: GeneratorSpec -> Edna ()
+addExperiments spec@GeneratorSpec {..} = do
+  forM_ [1 .. gsProjectNum] $ \projectIdx -> do
+    forM_ [1 .. gsMethodologyNum] $ \methodologyIdx ->
+      addExperiment projectIdx methodologyIdx
+    logMessage $ "Processed project " <> show projectIdx
+  where
+    addExperiment :: Word32 -> Word32 -> Edna ()
+    addExperiment projectId methodologyId = do
+      let description =
+            Text.replicate 100 (show projectId) <>
+            Text.replicate 100 (show methodologyId)
+      let metadata = replicate 100 (Text.replicate 1000 "x")
+      let fc = FileContents
+            { fcMeasurements = fileMeasurements spec projectId methodologyId
+            , fcMetadata = FileMetadata metadata
+            }
+      let blob = BSL.replicate 3e5 0x42
+      void $ Upload.uploadFile' (SqlId projectId) (SqlId methodologyId)
+        description "experiment.xlsx" blob fc
+
+fileMeasurements :: GeneratorSpec -> Word32 -> Word32 -> Map Text TargetMeasurements
+fileMeasurements GeneratorSpec {..} projectIdx methodologyIdx =
+  Map.fromList $ flip map [0 .. gsTargetNum - 1] $ \targetIdx ->
+    let
+      targetIdx' = (projectIdx - 1) * gsTargetNum + targetIdx
+      targetName = "TARGET number " <> show targetIdx'
+    in (targetName,) . TargetMeasurements . Map.fromList $
+      flip map [0 .. gsCompoundNum - 1] $ \compoundIdx ->
+        let
+          compoundIdx' = (projectIdx - 1) * gsCompoundNum + compoundIdx
+          compoundName = "COMPOUND number " <> show compoundIdx'
+        in (compoundName, measurements gsMeasurementsNum projectIdx methodologyIdx)
+
+measurements :: Word -> Word32 -> Word32 -> [Measurement]
+measurements n project methodology = map measurement [1 .. n]
+  where
+    measurement i =
+      let concentration = 200 / 2 ^ i in Measurement
+      { mConcentration = concentration
+      , mSignal =
+        realToFrac project / (1 + (concentration / realToFrac methodology)) +
```

Comment:
I think it'll be more readable if you add `signal` to let block, similar to `concentration`

Reply:
Done.